ADR/ADR004-store-form-schemas-in-an-api.md

```diff
@@ -0,0 +1,143 @@
+# ADR004: Store form schemas in an API
+
+Date: 2022-02-18
+
+## Status
+
+Proposed
+
+## Context
+
+- Manager/Designer components don't need to be as scalable as the runner
+- Runner needs to scale to number of users
+  - Sustainability, runner not needed during quiet times
+- "Published forms" need to be accessible to the runner
+
+### Option 1 - Forms API
+
+```mermaid
+graph TD
+  Manager[Form manager]
+
+  subgraph formsApiService[Forms API]
+    formsApi[API]
+    formStore[(Form datastore)]
+  end
+
+  runnerRenderer[Forms renderer]
+  Manager --> formsApi
+  formsApi --> formStore
+  formStore --> formsApi
+  runnerRenderer --> formsApi
+```
+
+- Stores the forms in a datastore with the API being a separate application
+- Runner depends on API
+- Three applications to deploy with dependencies, too complex?
+- Could cache API responses for speed
+
+### Option 2a - Runner as an API
+
+```mermaid
+flowchart
+  subgraph Manager
+    managerService[Form manager]
+    managerDatastore[(Form datastore)]
+  end
+
+  subgraph Runner[Runner application]
+    api[Publish API]
+    renderer[Forms renderer]
+    store[(Published form datastore)]
+  end
+
+  managerService --> api
+  renderer --> store
+  managerService --> managerDatastore
+  api --> store
+```
+
+- Stores the published forms alongside the runner
+- Runner doesn't depend on api but exposes it instead
+- Only two applications to deploy
+
+### Option 2b - Publishing API, stores in shared datastore
+
+```mermaid
+flowchart
+  subgraph Manager
+    managerService[Form manager]
+    managerDatastore[(Form datastore)]
+  end
+
+  subgraph SharedSpace[Shared space]
+    subgraph Runner[Runner application]
+      renderer[Forms renderer]
+    end
+
+    subgraph Publisher[Publishing application]
+      api[Publish API]
+    end
+    store[(Published form datastore)]
+  end
+
+  managerService --> api
+  renderer --> store
+  managerService --> managerDatastore
+  api --> store
+```
+
+- Stores the published forms in a DB directly accessible to the runner
+- Runner can be focused only on rendering
+- Publish API separate
+- Only two applications to deploy
+
+### Option 3 - Manager and Runner share database
+
+```mermaid
+flowchart
+  subgraph sharedSpace[GOV.UK Forms]
+    subgraph manager[Manager application]
+      managerService[Form manager]
+    end
+
+    formDatastore[(Form datastore)]
+
+    subgraph runner[Runner application]
+      formRenderer[Form renderer]
+    end
+  end
+
+  managerService <--> formDatastore
+  formDatastore --> formRenderer
+```
+
+- Form manager and runner share the same DB
+- No need to build/maintain APIs
+- Shared DB schema is a concern
+
+### Option 4 - Monolithic manager and runner
```

Comment:
In terms of development ease, a monolith is often the easiest one - We'd only have one service to deploy, one code base, dont need to manage dependencies between them etc etc. 

There's an argument I think to keeping a monolith but building it in a way that it's theoretically easy to progress towards separation - it'll be a lot easier and our volume is likely to be low enough for the first while that we could do that. But I wonder if that's creating problems down the line with anything unavoidable that could occur (I'm thinking of scale mainly)

Reply:
I agree that a monolith is easiest to develop quickly. Though, I really like the idea of the runner and api being separate and horizontally scalable, it seems an inevitable requirement?

Reply:
yeah I agree it feels like an inevitable requirement - normally I'd err on the side of "you can just scale the monolith that's fine" but with our platform, assuming we're successful, the runner aspect is going to face a lot more traffic than the management side and it feels inevitable we'd want to scale those applications differently